semconv/internal/pool/attr.go

```diff
@@ -0,0 +1,71 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package pool provides common pools for semantic convention implementations.
+package pool // import "go.opentelemetry.io/otel/semconv/internal/pool"
+
+import (
+	"slices"
+	"sync"
+
+	"go.opentelemetry.io/otel/attribute"
+)
+
+// MaxCapAttrsSmall is the maximum capacity of a slice that will be taken from
+// the small pool. Slices with a capacity greater than this will be taken from
+// the medium pool.
+const MaxCapAttrsSmall = 8
+
+var (
+	// attrSlice is a pool for slices of attribute.KeyValue that are less than
+	// or equal to the cutoffSmall.
+	attrSliceSmall = &sync.Pool{
+		New: func() any {
+			var s []attribute.KeyValue
+			// Return pointer to a slice instead of a slice to avoid an
+			// allocation when putting it back in the pool.
+			return &s
+		},
+	}
+
+	attrSliceMedium = &sync.Pool{
```

Comment:
Based on the test results from https://cloud-native.slack.com/archives/C0540L05T0X/p1756265527306699, it seems that returning a non-pointer here would yield better performance.